dali/python/nvidia/dali/_multiproc/shared_batch.py

```diff
@@ -217,42 +192,60 @@ def _add_array_to_batch(self, np_array, meta, memview):
-        for (idx, sample), (meta_idx, sample_meta) in zip(batch, meta):
-            assert idx == meta_idx
+        self.total_size = _align_up(self.data_size + self.meta_data_size, self.SAMPLE_ALIGNMENT)
+        if self.shm_chunk.capacity < self.total_size:
+            self.resize_shm_chunk(self.shm_chunk, self.total_size + self.min_trailing_offset)
+        memview = self.shm_chunk.buf
+        for sample, sample_meta in zip(batch, meta):
             _apply_to_sample(self._add_array_to_batch, sample, sample_meta, memview, nest_with_sample=1)
         # copy meta data at the end of shared memory chunk
         buffer = memview[self.data_size:(self.data_size + self.meta_data_size)]
         buffer[:] = serialized_meta
 
+    @staticmethod
+    def resize_shm_chunk(shm_chunk, needed_capacity):
+        new_capacity = max(needed_capacity, 2 * shm_chunk.capacity)
+        new_capacity = _align_up(new_capacity, SharedBatchWriter.BUFFER_ALIGNMENT)
+        shm_chunk.resize(new_capacity, trunc=True)
 
 
-def write_batch(mem_batch: SharedMemChunk, batch):
-    """Serialize and write the indexed data batch `batch` into the shared memory `mem_batch`.
-
-    Returns description of serialized memory.
-
-    Parameters
-    ----------
-    mem_batch : SharedMemChunk
-        Target memory to write to.
-    batch : List of (idx, Sample)
-        Batch of data to be serialized
-
-    Returns
-    -------
-        SharedBatchMeta
+class BufShmChunk:
+    """Simple wrapper around shared memory chunks. Adds mem_chunk_id used
+    to identify chunks in the communication between parent and worker process.
     """
-    sbw = SharedBatchWriter(mem_batch, batch)
-    return SharedBatchMeta.from_writer(sbw)
+    def __init__(self, shm_chunk_id, shm_chunk : shared_mem.SharedMem):
+        self.shm_chunk_id = shm_chunk_id
+        self.shm_chunk = shm_chunk
+
+    @classmethod
+    def allocate(cls, shm_chunk_id, initial_chunk_size):
+        return cls(shm_chunk_id, shared_mem.SharedMem.allocate(initial_chunk_size))
+
+
+def read_shm_message(shm_chunk : shared_mem.SharedMem, shm_message):
+    if shm_message.shm_capacity != shm_chunk.capacity:
+        shm_chunk.resize(shm_message.shm_capacity, trunc=False)
+    buffer = shm_chunk.buf[shm_message.offset:shm_message.offset+shm_message.num_bytes]
+    return pickle.loads(buffer)
+
+
+def write_shm_message(worker_id, shm : BufShmChunk, message, offset, resize=True):
+    serialized_message = pickle.dumps(message)
+    num_bytes = len(serialized_message)
+    if num_bytes > shm.shm_chunk.capacity - offset:
+        if resize:
+            SharedBatchWriter.resize_shm_chunk(shm.shm_chunk, offset + num_bytes)
+        else:
+            raise RuntimeError(
+                "Could not put message into shared memory region, not enough space in the buffer. "
+                "Consider specifying `bytes_per_minibatch_hint` of parallel external source to at "
+                "least {}".format(offset + num_bytes))
```

Comment:
I don't see much value in that message.
I think it should always be resized. Because the fact the user adjusts the value to `offset + num_bytes` doesn't give any guarantee it will work for the next chunk/batch/message.

Reply:
This would happen only if parent process tried to put task description that doesn't fit into the given chunk. They are initialized to 1MB and resized by the worker if actual batch needs more space. I thought it is rather unusual situation to have such a big task description (it would have to be sample mode with batch size so big that mere SampleInfo list exceeds that size). Additionally I assumed that when workers resize the chunk for data then it later on it is even more unlikely for the task description to take more space than actual data.

If, however, that would be the case, user could specify bytes_per_minibatch_hint to be big enough.

Parent process could resize that if needed but then I need to drop the idea of closing file descriptors for shm in parent process.

Reply:
Ok. Up to you then.